router-plugin/tracing/tracer.go

```diff
@@ -0,0 +1,198 @@
+
+	DefaultBatchTimeout  = 10 * time.Second
+	DefaultExportTimeout = 30 * time.Second
+
+	SampleRate = 1.0
+)
+
+type OpenTelemetryExporter struct {
+	Endpoint string
+
+	Exporter      Exporter
+	BatchTimeout  time.Duration
+	ExportTimeout time.Duration
+	Headers       map[string]string
+	HTTPPath      string
+}
+
+func initTracer(ctx context.Context, optParams TracingOptions) (*sdktrace.TracerProvider, error) {
+	var exporters []*OpenTelemetryExporter
+	if exporterConfig := os.Getenv("router_exporter_config"); exporterConfig != "" {
+		err := json.Unmarshal([]byte(exporterConfig), &exporters)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Return no-op provider
+	if len(exporters) == 0 {
+		provider := sdktrace.NewTracerProvider(sdktrace.WithSampler(sdktrace.NeverSample()))
+		otel.SetTracerProvider(provider)
+		return provider, nil
+	}
+
+	r, err := resource.New(ctx,
+		resource.WithAttributes(semconv.ServiceNameKey.String(optParams.ServiceName)),
+		resource.WithAttributes(semconv.ServiceVersionKey.String(optParams.ServiceVersion)),
+		resource.WithAttributes(attribute.Key("wg.is_plugin").Bool(true)),
+		resource.WithProcessPID(),
+		resource.WithOSType(),
+		resource.WithTelemetrySDK(),
+		resource.WithHost(),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	opts := []sdktrace.TracerProviderOption{
+		sdktrace.WithRawSpanLimits(sdktrace.SpanLimits{
+			AttributeValueLengthLimit:   3 * 1024, // 3KB
+			AttributeCountLimit:         sdktrace.DefaultAttributeCountLimit,
+			EventCountLimit:             sdktrace.DefaultEventCountLimit,
+			LinkCountLimit:              sdktrace.DefaultLinkCountLimit,
+			AttributePerEventCountLimit: sdktrace.DefaultEventCountLimit,
+			AttributePerLinkCountLimit:  sdktrace.DefaultAttributePerLinkCountLimit,
+		}),
+		sdktrace.WithResource(r),
+	}
+
+	opts = append(opts,
+		sdktrace.WithSampler(
+			sdktrace.ParentBased(
+				sdktrace.TraceIDRatioBased(SampleRate),
+			),
+		),
+	)
+
+	for _, exp := range exporters {
+		// Default to OLTP HTTP
+		if exp.Exporter == "" {
+			exp.Exporter = ExporterOLTPHTTP
+		}
+
+		exporter, err := createExporter(exp)
+		if err != nil {
+			return nil, err
+		}
+
+		batchTimeout := exp.BatchTimeout
+		if batchTimeout == 0 {
+			batchTimeout = DefaultBatchTimeout
+		}
+
+		exportTimeout := exp.ExportTimeout
+		if exportTimeout == 0 {
+			exportTimeout = DefaultExportTimeout
+		}
+
+		// Always be sure to batch in production.
+		opts = append(opts,
+			sdktrace.WithBatcher(exporter,
+				sdktrace.WithBatchTimeout(batchTimeout),
+				sdktrace.WithExportTimeout(exportTimeout),
+				sdktrace.WithMaxExportBatchSize(512),
+				sdktrace.WithMaxQueueSize(2048),
+			),
+		)
+	}
+
+	// Set global tracer provider
+	tp := sdktrace.NewTracerProvider(opts...)
+
+	otel.SetTracerProvider(tp)
+
+	// Set global propagator for trace context propagation
+	otel.SetTextMapPropagator(
+		propagation.NewCompositeTextMapPropagator(propagation.TraceContext{}, propagation.Baggage{}))
+
+	return tp, nil
```

Comment:
We currently support these propagators in the router config:
https://cosmo-docs.wundergraph.com/router/open-telemetry#trace-propagation